### PR TITLE
[v2] Configure health check extension for all configs

### PR DESCRIPTION
**Pull request by Wise-Wizard**
**Which problem is this PR solving?**

Part of #5633, part of #5859

**Description of the changes**
* Integrate health check extension to monitor and report Jaeger V2 component's health
* Enhance all-in-one CI test to ping the new health port

**How was this change tested?**

The changes were tested by running the following command:

```bash
make test
```
```bash
CI actions and new Unit Tests
```
**Checklist**

- [x] I have read [CONTRIBUTING_GUIDELINES.md](https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md)
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - `for jaeger: make lint test`
  - `for jaeger-ui: yarn lint` and `yarn test`

### CODE REVIEW

**Comment by Wise-Wizard** [comment]
HI, so the health check configuration should be added to every config file individually itself right?

**Comment by yurishkuro** [comment]
Yes. And you need your mention it in the service section.

**Comment by yurishkuro** [comment]
See https://github.com/jaegertracing/jaeger/issues/5859

**Comment by Wise-Wizard** [comment]
> See #5859

Step 1 is done, I will proceed with e2e tests now.

File: cmd/jaeger/collector-with-kafka.yaml [hunk 1]
```diff
@@ -1,4 +1,5 @@
 service:
```
Comment (yurishkuro) [inline cmd/jaeger/collector-with-kafka.yaml:1]: please rename this file to `config-kafka-collector.yaml`

File: cmd/jaeger/ingester-remote-storage.yaml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 service:
```
Comment (yurishkuro) [inline cmd/jaeger/ingester-remote-storage.yaml:1]: please rename this file `config-kafka-ingester.yaml`

File: cmd/jaeger/ingester-remote-storage.yaml [hunk 3]
```diff
@@ -12,6 +12,11 @@ service:
       level: debug
 
 extensions:
+  healthcheckv2:
+    use_v2: true
+    http:
+      endpoint: "0.0.0.0:13133"
```
Comment (yurishkuro) [inline cmd/jaeger/ingester-remote-storage.yaml:18]: will you not get a port conflict if both collector and ingester use the same port for HC extension? E.g. we bumped the metrics port from 8888 to 8889 above.

File: cmd/jaeger/internal/integration/e2e_integration.go [hunk 4]
```diff
@@ -109,7 +110,24 @@ func (s *E2EStorageIntegration) e2eInitialize(t *testing.T, storage string) {
 			return false
 		}
 		defer resp.Body.Close()
-		return resp.StatusCode == http.StatusOK
+		if s.HealthCheckEndpoint == "" {
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/e2e_integration.go:113]: this is never true, see L96

On a separate question, is `ports.QueryHTTP` what you want to be checking (for non-Kafka tests)?

Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: ports.QueryHTTP was already written in the code before by Harshit.
Didn't bother to tinker with that.

Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: L96 is initializing the healthCheckEndpoint, but the s.HealthCheckEndpoint if "" still remains empty even after healthCheckEndpoint is initialized if I am right?

Comment (yurishkuro) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: what is the possible execution path for this variable to remain empty after this code?
```
	healthCheckEndpoint := s.HealthCheckEndpoint
	if healthCheckEndpoint == "" {
		healthCheckEndpoint = fmt.Sprintf("http://localhost:%d/", ports.QueryHTTP)
	}
```

Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: healthCheckEndpoint will not be empty.
But, the condition tests for if s.HealthCheckEndpoint is empty, and s.HealthCheckEndpoint will always be empty from where it is initialized.

Comment (yurishkuro) [inline cmd/jaeger/internal/integration/e2e_integration.go:113]: We don't want this branching. All e2e tests now run the health extension, why do we need to special case?

Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: Noted, will make the change.

Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: I am quite not able to figure out, as to why the endpoint is saying connection refused for Kafka. It doesn't say the same for other config files?

Comment (yurishkuro) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: I don't think Kafka is starting, why would it have no logs?
image

Comment (yurishkuro) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: actually this might be expected, since we do shutdown in the script
```
teardown_kafka() {
  echo "Stopping Kafka..."
  docker compose -f "${compose_file}" down
}
```
this is a bug - in other scripts we dump the logs from teardown, before we shutdown the compose env.

Comment (yurishkuro) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: another thing I do not see is the output of the logs from the two binaries you're starting, why is that?

Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: > actually this might be expected, since we do shutdown in the script
> 
> ```
> teardown_kafka() {
>   echo "Stopping Kafka..."
>   docker compose -f "${compose_file}" down
> }
> ```
> 
> this is a bug - in other scripts we dump the logs from teardown, before we shutdown the compose env.

@joeyyy09 PTAL at this.

Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: > another thing I do not see is the output of the logs from the two binaries you're starting, why is that?

I don't think I have made any changes that should affect the logs at any level. I have just changed the healthcheck URL.
That should have no effect whatsoever.

Comment (yurishkuro) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: I am not saying it's because of your changes, I am saying that we do have code in e2e struct that dumps the logs of the binaries, and I do not see this happening for Kafka tests. So this is both an operational issue and a blocker for you to troubleshoot why the health checks are failing in the test.

Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:113]: Ohhh Okay.

File: cmd/jaeger/internal/integration/e2e_integration.go [hunk 5]
```diff
@@ -109,7 +110,21 @@ func (s *E2EStorageIntegration) e2eInitialize(t *testing.T, storage string) {
 			return false
 		}
 		defer resp.Body.Close()
-		return resp.StatusCode == http.StatusOK
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/e2e_integration.go:112]: may want to keep this as early exit condition `if resp.StatusCode != http.StatusOK`

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 6]
```diff
@@ -97,19 +97,20 @@ func TestKafkaStorage(t *testing.T) {
 
 			collector := &E2EStorageIntegration{
 				SkipStorageCleaner:  true,
-				ConfigFile:          createConfigWithEncoding(t, "../../collector-with-kafka.yaml", test.encoding, uniqueTopic),
-				HealthCheckEndpoint: "http://localhost:8888/metrics",
+				ConfigFile:          createConfigWithEncoding(t, "../../config-kafka-collector.yaml", test.encoding, uniqueTopic),
+				HealthCheckEndpoint: "http://localhost:13133/status",
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:101]:
```suggestion
```
this is the default

File: cmd/jaeger/internal/integration/e2e_integration.go [hunk 7]
```diff
@@ -88,30 +89,6 @@ func (s *E2EStorageIntegration) e2eInitialize(t *testing.T, storage string) {
...
-		}
-		resp, err := http.DefaultClient.Do(req)
-		if err != nil {
-			t.Logf("HTTP request failed: %v", err)
-			return false
-		}
-		defer resp.Body.Close()
-		return resp.StatusCode == http.StatusOK
-	}, 60*time.Second, 3*time.Second, "Jaeger-v2 did not start")
-	t.Log("Jaeger-v2 is ready")
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/e2e_integration.go:114]: I moved the healthcheck **after** the call to Cleanup, that was the reason why the binary's logs were never printed (`require` it the HC would abort the test).

**Comment by yurishkuro** [comment]
error in the logs
```
  2024-08-21T21:59:58.834Z	error	extensions/extensions.go:52	Failed to start extension	{"kind": "extension", "name": "healthcheckv2", "error": "listen tcp 127.0.0.1:13132: bind: address already in use"}
```

File: cmd/jaeger/config-kafka-ingester.yaml [hunk 8]
```diff
@@ -12,6 +12,12 @@ service:
       level: debug
 
 extensions:
+  healthcheckv2:
+    use_v2: true
+    http:
+      endpoint: 0.0.0.0:14133
+    grpc:
+      endpoint: 0.0.0.0:14132
```
Comment (yurishkuro) [inline cmd/jaeger/config-kafka-ingester.yaml:20]: ingester is already using non-standard metrics port, so I changed it to use non-standard health check ports too

**Comment by yurishkuro** [comment]
kafka tests are green, but grpc v2 is consistently failing. It's also missing logs from one of the binaries.

File: cmd/jaeger/internal/integration/e2e_integration.go (hunk 7 above)
Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:114]: Ohh okay.

**Comment by Wise-Wizard** [comment]
> kafka tests are green, but grpc v2 is consistently failing. It's also missing logs from one of the binaries.

image
This is the error that is being shown.
Does this have to do something with the healthcheck extension?

File: cmd/jaeger/internal/integration/e2e_integration.go [hunk 9]
```diff
@@ -144,12 +126,67 @@ func (s *E2EStorageIntegration) e2eInitialize(t *testing.T, storage string) {
 	s.SpanReader, err = createSpanReader(logger, ports.QueryGRPC)
 	require.NoError(t, err)
 
+	// Wait for the binary to start and become ready to serve requests.
+	require.Eventually(t, func() bool { return s.doHealthCheck(t) },
+		60*time.Second, 3*time.Second, "%s did not start", s.BinaryName)
+	t.Logf("%s is ready", s.BinaryName)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/e2e_integration.go:132]: Why did you move it here? It is again breaking the logical flow and the cleanup order.

Comment (Wise-Wizard) [reply cmd/jaeger/internal/integration/e2e_integration.go:132]: Yeah, moving it back. Was just testing.

**Comment by yurishkuro** [comment]
While the test is running you can query it for traces with jaeger service name to see which endpoint is causing it to appear. It's possible that it comes from the health extension, although it would be strange since we did not enable global tracer, only jtracer for our parts.

**Comment by Wise-Wizard** [comment]
Hi @yurishkuro, So I ran the gRPC integration test locally and the logs are perfect and no error is being reported in the test.
The jaeger variable is not being added in getService operation when running the storage-integration-test locally.
image
image

**Comment by yurishkuro** [comment]
I added logging of the traces corresponding to the service names that are retrieved, and turns out the "offending" service name "jaeger" is coming from traces on the write path
```
    integration.go:164: span: Service: jaeger, TraceID: 0e5d9afa746db440f417202ece817ba7, 
            Operation: jaeger.storage.v1.SpanWriterPlugin/WriteSpan
```
This is actually (a) a great find, and (b) a problem, because we should never be tracing the write path due to the risk of infinite self-trace loop.

**Comment by yurishkuro** [comment]
what is weird is why this doesn't seem to happen when running the test locally...

**Comment by Wise-Wizard** [comment]
Yes exactly, I had added the same logs in order to debug it, but surprisingly locally all the tests were being passed.

**Comment by yurishkuro** [comment]
It's a timing issue. The traces used in the unit tests are directly written to the pipeline and quickly end up in the storage. But the extraneous traces from "jaeger" first go through SDK's batch exporter, which buffers spans for some time before flushing. In local tests those "jaeger" traces don't make it in time for the test traces to be detected. I was able to reproduce it locally with this diff:
```diff
diff --git a/cmd/jaeger/internal/integration/e2e_integration.go b/cmd/jaeger/internal/integration/e2e_integration.go
index 55f1517f..c65ca281 100644
--- a/cmd/jaeger/internal/integration/e2e_integration.go
+++ b/cmd/jaeger/internal/integration/e2e_integration.go
@@ -184,6 +184,8 @@ func (s *E2EStorageIntegration) doHealthCheck(t *testing.T) bool {
 		t.Logf("Received non-K status %s: %s", healthResponse.Status, string(body))
 		return false
 	}
+	t.Logf("%s is ready, but let's sleep a bit", s.BinaryName)
+	time.Sleep(30*time.Second + 3*time.Second)
 	return true
 }

diff --git a/cmd/jaeger/internal/integration/grpc_test.go b/cmd/jaeger/internal/integration/grpc_test.go
index 4f7c1cee..6cf0cddf 100644
--- a/cmd/jaeger/internal/integration/grpc_test.go
+++ b/cmd/jaeger/internal/integration/grpc_test.go
@@ -4,11 +4,9 @@
 package integration

 import (
-	"fmt"
 	"testing"

 	"github.com/jaegertracing/jaeger/plugin/storage/integration"
-	"github.com/jaegertracing/jaeger/ports"
 )

 type GRPCStorageIntegration struct {
@@ -33,7 +31,7 @@ func TestGRPCStorage(t *testing.T) {
 		E2EStorageIntegration: E2EStorageIntegration{
 			ConfigFile: "../../config-remote-storage.yaml",
 			// TODO this should be removed in favor of default health check endpoint
-			HealthCheckEndpoint: fmt.Sprintf("http://localhost:%d/", ports.QueryHTTP),
+			// HealthCheckEndpoint: fmt.Sprintf("http://localhost:%d/", ports.QueryHTTP),
 		},
 	}
 	s.CleanUp = s.cleanUp
```

**Comment by Wise-Wizard** [comment]
> It's a timing issue. The traces used in the unit tests are directly written to the pipeline and quickly end up in the storage. But the extraneous traces from "jaeger" first go through SDK's batch exporter, which buffers spans for some time before flushing. In local tests those "jaeger" traces don't make it in time for the test traces to be detected. I was able to reproduce it locally with this diff:
> 
> ```diff
> diff --git a/cmd/jaeger/internal/integration/e2e_integration.go b/cmd/jaeger/internal/integration/e2e_integration.go
> index 55f1517f..c65ca281 100644
> --- a/cmd/jaeger/internal/integration/e2e_integration.go
> +++ b/cmd/jaeger/internal/integration/e2e_integration.go
> @@ -184,6 +184,8 @@ func (s *E2EStorageIntegration) doHealthCheck(t *testing.T) bool {
>  		t.Logf("Received non-K status %s: %s", healthResponse.Status, string(body))
>  		return false
>  	}
> +	t.Logf("%s is ready, but let's sleep a bit", s.BinaryName)
> +	time.Sleep(30*time.Second + 3*time.Second)
>  	return true
>  }
> 
> diff --git a/cmd/jaeger/internal/integration/grpc_test.go b/cmd/jaeger/internal/integration/grpc_test.go
> index 4f7c1cee..6cf0cddf 100644
> --- a/cmd/jaeger/internal/integration/grpc_test.go
> +++ b/cmd/jaeger/internal/integration/grpc_test.go
> @@ -4,11 +4,9 @@
>  package integration
> 
>  import (
> -	"fmt"
>  	"testing"
> 
>  	"github.com/jaegertracing/jaeger/plugin/storage/integration"
> -	"github.com/jaegertracing/jaeger/ports"
>  )
> 
>  type GRPCStorageIntegration struct {
> @@ -33,7 +31,7 @@ func TestGRPCStorage(t *testing.T) {
>  		E2EStorageIntegration: E2EStorageIntegration{
>  			ConfigFile: "../../config-remote-storage.yaml",
>  			// TODO this should be removed in favor of default health check endpoint
> -			HealthCheckEndpoint: fmt.Sprintf("http://localhost:%d/", ports.QueryHTTP),
>  		},
>  	}
>  	s.CleanUp = s.cleanUp
> ```

Ah, makes sense now.
